File: codes/visual_basic_dotnet/array_reduce.vb
Imports System.Collections.Generic ' Dictionary, KeyValuePair

Module Program
    Function JsonStringify(ByVal Anything As Object, Optional ByVal Pretty As Boolean = False, Optional ByVal Indent As String = "    ") As String
        Dim IndentLevel As Integer = 0
        Dim JsonStringifyInner As Func(Of Object, String, String) = Function(ByVal AnythingInner As Object, ByVal IndentInner As String)
            If (AnythingInner Is Nothing) Then Return "null"
            If (TypeOf AnythingInner Is String) Then Return """" & AnythingInner & """"
            If (IsNumeric(AnythingInner) = True) Then Return CStr(AnythingInner).Replace(",", ".")
            If (TypeOf AnythingInner Is Boolean) Then Return CStr(AnythingInner)
            If (TypeOf AnythingInner Is List(Of Object)) Then
                If (AnythingInner.Count = 0) Then Return "[]"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "[" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "[")
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In AnythingInner
                    Result &= JsonStringifyInner(ArrayItem, IndentInner)
                    If ((ArrayItemIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ArrayItemIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "]", "]")
                Return Result
            End If
            If (TypeOf AnythingInner Is Dictionary(Of String, Object)) Then
                If (AnythingInner.Count = 0) Then Return "{}"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "{" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "{")
                Dim ObjectIterationIndex As Integer = 0
                For Each ObjectEntry As KeyValuePair(Of String, Object) In AnythingInner
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    Result &= """" & ObjectKey & """: " & JsonStringifyInner(ObjectValue, IndentInner)
                    If ((ObjectIterationIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ObjectIterationIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "}", "}")
                Return Result
            End If
            Return "null"
        End Function
        Return JsonStringifyInner(Anything, Indent)
    End Function

    Function SpreadObject(ByVal ParamArray Parameters() As Object) As Dictionary(Of String, Object)
        Dim NewObject As New Dictionary(Of String, Object)()
        Dim ParameterIndex As Integer = 0
        For Each Parameter As Object In Parameters
            If (TypeOf Parameter Is Dictionary(Of String, Object)) Then
                For Each ObjectEntry As KeyValuePair(Of String, Object) In Parameter
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    NewObject(CStr(ObjectKey)) = ObjectValue
                Next
                Continue For
            End If
            If (TypeOf Parameter Is List(Of Object)) Then
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In Parameter
                    NewObject(CStr(ArrayItemIndex)) = ArrayItem
                    ArrayItemIndex += 1
                Next
                Continue For
            End If
            ParameterIndex += 1
        Next
        Return NewObject
    End Function

    Function SpreadArray(ByVal ParamArray Parameters() As Object) As List(Of Object)
        Dim NewArray As New List(Of Object)()
        Dim ParameterIndex As Integer = 0
        For Each Parameter As Object In Parameters
            If (TypeOf Parameter Is Dictionary(Of String, Object)) Then
                If (Parameter.Count = 1) Then
                    For Each ObjectEntry As KeyValuePair(Of String, Object) In Parameter
                        Dim ObjectKey = ObjectEntry.Key
                        Dim ObjectValue = ObjectEntry.Value
                        NewArray.Add(ObjectValue)
                    Next
                    Continue For
                End If
                NewArray.Add(Parameter)
                Continue For
            End If
            If (TypeOf Parameter Is List(Of Object)) Then
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In Parameter
                    NewArray.Add(ArrayItem)
                    ArrayItemIndex += 1
                Next
                Continue For
            End If
            ParameterIndex += 1
        Next
        Return NewArray
    End Function

    Function ArrayReduce(ByVal CallbackFunction As Func(Of Object, Object, Integer, List(Of Object), Object), ByVal AnArray As List(Of Object), ByVal InitialValue As Object) As Object
        ' JavaScript-like Array.reduce() function
        Dim Result As Object = InitialValue
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            Result = CallbackFunction(Result, ArrayItem, ArrayItemIndex, AnArray)
            ArrayItemIndex += 1
        Next
        Return Result
    End Function

    Sub Main(Args As String())
        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.reduce() in Visual Basic (.NET) List")

        Dim Numbers As List(Of Object) = New List(Of Object) From {36, 57, 2.7, 2.3, -12, -34, -6.5, -4.3}
        Console.WriteLine($"Numbers: {JsonStringify(Numbers)}")

        Dim NumbersTotal As Double = 0.0

        Console.WriteLine("' using JavaScript-like Array.reduce() function ""ArrayReduce""")

        NumbersTotal = ArrayReduce(Function(ByVal CurrentResult As Object, ByVal CurrentNumber As Double, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (CurrentResult + CurrentNumber), Numbers, 0.0)
        Console.WriteLine($"total number: {JsonStringify(NumbersTotal)}")
        ' total number: 41.2

        Console.WriteLine("' using Visual Basic (.NET) Array.reduce() built-in method ""Array.Aggregate()""")

        NumbersTotal = Numbers.Aggregate(0.0, Function(ByVal CurrentResult As Object, ByVal CurrentNumber As Double) (CurrentResult + CurrentNumber))
        Console.WriteLine($"total number: {JsonStringify(NumbersTotal)}")
        ' total number: 41.2

        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.reduce() in Visual Basic (.NET) List of Dictionaries")

        Dim Products As List(Of Object) = New List(Of Object) From {
            New Dictionary(Of String, Object) From {
                {"code", "pasta"},
                {"price", 321}
            },
            New Dictionary(Of String, Object) From {
                {"code", "bubble_gum"},
                {"price", 233}
            },
            New Dictionary(Of String, Object) From {
                {"code", "potato_chips"},
                {"price", 5}
            },
            New Dictionary(Of String, Object) From {
                {"code", "towel"},
                {"price", 499}
            }
        }
        Console.WriteLine($"Products: {JsonStringify(Products, Pretty:=True)}")

        Dim ProductsGrouped As Object

        Console.WriteLine("' using JavaScript-like Array.reduce() function ""ArrayReduce""")

        ProductsGrouped = ArrayReduce(Function(ByVal CurrentResult As Object, ByVal CurrentProduct As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) If((CurrentProduct("price") > 100), SpreadObject(CurrentResult, New Dictionary(Of String, Object) From {{"expensive", SpreadArray(CurrentResult("expensive"), New Dictionary(Of String, Object) From {{"CurrentProduct", CurrentProduct}})}}), SpreadObject(CurrentResult, New Dictionary(Of String, Object) From {{"cheap", SpreadArray(CurrentResult("cheap"), New Dictionary(Of String, Object) From {{"CurrentProduct", CurrentProduct}})}})), Products, New Dictionary(Of String, Object) From {{"expensive", New List(Of Object)}, {"cheap", New List(Of Object)}})
        Console.WriteLine($"grouped products: {JsonStringify(ProductsGrouped, Pretty:=True)}")
        ' grouped products: {
        '     "expensive": [
        '         {
        '             "code": "pasta",
        '             "price": 321
        '         },
        '         {
        '             "code": "bubble_gum",
        '             "price": 233
        '         },
        '         {
        '             "code": "towel",
        '             "price": 499
        '         }
        '     ],
        '     "cheap": [
        '         {
        '             "code": "potato_chips",
        '             "price": 5
        '         }
        '     ]
        ' }

        Console.WriteLine("' using Visual Basic (.NET) Array.reduce() built-in method ""Array.Aggregate()""")

        ProductsGrouped = Products.Aggregate(New Dictionary(Of String, Object) From {{"expensive", New List(Of Object)}, {"cheap", New List(Of Object)}}, Function(ByVal CurrentResult As Object, ByVal CurrentProduct As Object) If((CurrentProduct("price") > 100), SpreadObject(CurrentResult, New Dictionary(Of String, Object) From {{"expensive", SpreadArray(CurrentResult("expensive"), New Dictionary(Of String, Object) From {{"CurrentProduct", CurrentProduct}})}}), SpreadObject(CurrentResult, New Dictionary(Of String, Object) From {{"cheap", SpreadArray(CurrentResult("cheap"), New Dictionary(Of String, Object) From {{"CurrentProduct", CurrentProduct}})}})))
        Console.WriteLine($"grouped products: {JsonStringify(ProductsGrouped, Pretty:=True)}")
        ' grouped products: {
        '     "expensive": [
        '         {
        '             "code": "pasta",
        '             "price": 321
        '         },
        '         {
        '             "code": "bubble_gum",
        '             "price": 233
        '         },
        '         {
        '             "code": "towel",
        '             "price": 499
        '         }
        '     ],
        '     "cheap": [
        '         {
        '             "code": "potato_chips",
        '             "price": 5
        '         }
        '     ]
        ' }
    End Sub
End Module

File: codes/gnu_octave/array_reduce.m
function stringrepeatresult = stringrepeat(astring, count)
    result = "";
    for i = (1:1:count) % (start:step:stop)
        result = cstrcat(result, astring);
    end
    stringrepeatresult = result;
end

function arrayreduceresult = arrayreduce(callbackfunction, anarray, initialvalue)
    % JavaScript-like Array.reduce() function
    currentresult = initialvalue;
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        currentresult = callbackfunction(currentresult, arrayitem, arrayitemindex, anarray);
    end
    arrayreduceresult = currentresult;
end

function optionalchainingresult = optionalchaining(anything, varargin)
    % JavaScript-like Optional Chaining Operator (?.) function
    % varargin == objectpropertiesarray
    if ((~isstruct(anything) && ~iscell(anything)) || isempty(varargin))
        optionalchainingresult = anything;
        return;
    end
    function arrayreducecallbackresult = arrayreducecallback(currentresult, currentitem, varargin)
        % varargin == (arrayitemindex, anarray)
        if ((isempty(currentresult) && isstruct(anything) && (ischar(currentitem) || ischar(currentitem))) && isfield(anything, currentitem))
            arrayreducecallbackresult = anything.(currentitem);
            return;
        end
        if (isempty(currentresult) && iscell(anything) && isnumeric(currentitem) && (currentitem >= 1) && (numel(anything) > currentitem))
            arrayreducecallbackresult = anything{currentitem};
            return;
        end
        if ((isstruct(currentresult) && (ischar(currentitem) || ischar(currentitem))) && isfield(currentresult, currentitem))
            arrayreducecallbackresult = currentresult.(currentitem);
            return;
        end
        if (iscell(currentresult) && isnumeric(currentitem) && (currentitem >= 1) && (numel(currentresult) > currentitem))
            arrayreducecallbackresult = currentresult{currentitem};
            return;
        end
        arrayreducecallbackresult = {};
    end
    optionalchainingresult = arrayreduce(@arrayreducecallback, varargin, {});
end

function ternaryresult = ternary(truecondition, callbackfunctionifconditiontrue, callbackfunctionifconditionfalse)
    if (islogical(truecondition) && (truecondition == true))
        ternaryresult = callbackfunctionifconditiontrue();
        return;
    end
    ternaryresult = callbackfunctionifconditionfalse();
end

function jsonstringifyresult = jsonstringify(anything, varargin)
    indentlevel = 0;
    function jsonstringifyinnerresult = jsonstringifyinner(anythinginner, prettyinner, indentinner)
        if isempty(anythinginner)
            jsonstringifyinnerresult = "null";
            return;
        end
        if (ischar(anythinginner) || ischar(anythinginner))
            jsonstringifyinnerresult = cstrcat("""", anythinginner, """");
            return;
        end
        if isnumeric(anythinginner)
            jsonstringifyinnerresult = num2str(anythinginner);
            return;
        end
        if (islogical(anythinginner) && (anythinginner == true))
            jsonstringifyinnerresult = "true";
            return;
        end
        if (islogical(anythinginner) && (anythinginner == false))
            jsonstringifyinnerresult = "false";
            return;
        end
        if iscell(anythinginner)
            if isempty(anythinginner)
                jsonstringifyinnerresult = "[]";
                return;
            end
            indentlevel = indentlevel + 1;
            result = ternary((pretty == true), @() (cstrcat("[", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() ("["));
            for arrayitemindex = (1:1:numel(anythinginner)) % (start:step:stop)
                arrayitem = anythinginner{arrayitemindex};
                result = cstrcat(result, jsonstringifyinner(arrayitem, prettyinner, indentinner));
                if (arrayitemindex ~= numel(anythinginner))
                    result = ternary((pretty == true), @() (cstrcat(result, ",", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() (cstrcat(result, ", ")));
                end
            end
            indentlevel = indentlevel - 1;
            result = ternary((pretty == true), @() (cstrcat(result, sprintf("\n"), stringrepeat(indentinner, indentlevel), "]")), @() (cstrcat(result, "]")));
            jsonstringifyinnerresult = result;
            return;
        end
        if isstruct(anythinginner)
            if (numel(fieldnames(anythinginner)) == 0)
                jsonstringifyinnerresult = "{}";
                return;
            end
            indentlevel = indentlevel + 1;
            result = ternary((pretty == true), @() (cstrcat("{", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() ("{"));
            objectkeys = fieldnames(anythinginner);
            for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                objectkey = objectkeys{objectentryindex};
                objectvalue = anythinginner.(objectkey);
                result = cstrcat(result, """", objectkey, """: ", jsonstringifyinner(objectvalue, prettyinner, indentinner));
                if (objectentryindex ~= numel(objectkeys))
                    result = ternary((pretty == true), @() (cstrcat(result, ",", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() (cstrcat(result, ", ")));
                end
            end
            indentlevel = indentlevel - 1;
            result = ternary((pretty == true), @() (cstrcat(result, sprintf("\n"), stringrepeat(indentinner, indentlevel), "}")), @() (cstrcat(result, "}")));
            jsonstringifyinnerresult = result;
            return;
        end
        jsonstringifyinnerresult = "null";
    end
    prettydefault = false;
    indentdefault = "    ";
    pretty = prettydefault;
    indent = indentdefault;
    if isempty(varargin)
        jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
        return;
    end
    optionalargument = varargin{1};
    if (islogical(optionalargument) && (optionalargument == false))
        jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
        return;
    end
    if isstruct(optionalargument)
        pretty = optionalchaining(optionalargument, "pretty");
        indent = optionalchaining(optionalargument, "indent");
        pretty = ternary(isempty(pretty), @() (prettydefault), @() (pretty));
        indent = ternary(isempty(indent), @() (indentdefault), @() (indent));
        jsonstringifyresult = jsonstringifyinner(anything, pretty, indent);
        return;
    end
    if (islogical(optionalargument) && (optionalargument == true))
        if (numel(varargin) >= 2)
            optionalargument2 = varargin{2};
            if (ischar(optionalargument2) || ischar(optionalargument2))
                indent = optionalargument2
            end
        end
        pretty = optionalargument;
        jsonstringifyresult = jsonstringifyinner(anything, pretty, indent);
        return;
    end
    jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
end

function sprint(varargin)
    currentresult = "";
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if (iscell(argument) && (numel(argument) == 1))
            argumentnew = jsonstringify(argument{1});
            argumentnew = strrep(argumentnew, """{", "{");
            argumentnew = strrep(argumentnew, """[", "[");
            argumentnew = strrep(argumentnew, "}""", "}");
            argumentnew = strrep(argumentnew, "]""", "]");
            currentresult = cstrcat(currentresult, argumentnew);
            continue;
        end
        if (~ischar(argument) && ~ischar(argument))
            error("Non string argument must be wrapped in {}");
            continue;
        end
        currentresult = cstrcat(currentresult, argument);
    end
    disp(currentresult);
end

function spreadobjectresult = spreadobject(varargin)
    newobject = struct();
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if isstruct(argument)
            objectkeys = fieldnames(argument);
            for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                objectkey = objectkeys{objectentryindex};
                objectvalue = argument.(objectkey);
                newobject.(objectkey) = objectvalue;
            end
            continue;
        end
        if iscell(argument)
            for arrayitemindex = (1:1:numel(argument)) % (start:step:stop)
                arrayitem = argument{arrayitemindex};
                newobject.(cstrcat("index", num2str(arrayitemindex))) = arrayitem;
            end
            continue;
        end
    end
    spreadobjectresult = newobject;
end

function spreadarrayresult = spreadarray(varargin)
    newarray = {};
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if isstruct(argument)
            objectkeys = fieldnames(argument);
            if (numel(objectkeys) == 1)
                for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                    objectkey = objectkeys{objectentryindex};
                    objectvalue = argument.(objectkey);
                    newarray{end + 1} = objectvalue;
                end
                continue;
            end
            newarray{end + 1} = argument;
            continue;
        end
        if iscell(argument)
            for arrayitemindex = (1:1:numel(argument)) % (start:step:stop)
                arrayitem = argument{arrayitemindex};
                newarray{end + 1} = arrayitem;
            end
            continue;
        end
    end
    spreadarrayresult = newarray;
end

sprint(sprintf("\n"), "% JavaScript-like Array.reduce() in GNU Octave Cell-Array");

numbers = {36, 57, 2.7, 2.3, -12, -34, -6.5, -4.3};
sprint("numbers: ", jsonstringify(numbers));

sprint("% using JavaScript-like Array.reduce() function ""arrayreduce""");

numberstotal = arrayreduce(@(currentresult, currentnumber, varargin) (currentresult + currentnumber), numbers, 0);
sprint("total number: ", {numberstotal});
% total number: 41.2

sprint(sprintf("\n"), "% JavaScript-like Array.reduce() in GNU Octave Cell-Array of structs");

products = { ...
    struct( ...
        "code", {"pasta"}, ...
        "price", {321} ...
    ), ...
    struct( ...
        "code", {"bubble_gum"}, ...
        "price", {233} ...
    ), ...
    struct( ...
        "code", {"potato_chips"}, ...
        "price", {5} ...
    ), ...
    struct( ...
        "code", {"towel"}, ...
        "price", {499} ...
    ) ...
};
sprint("products: ", jsonstringify(products, struct("pretty", {true})));

sprint("% using JavaScript-like Array.reduce() function ""arrayreduce""");

productsgrouped = arrayreduce(@(currentresult, currentproduct, varargin) ternary((currentproduct.("price") > 100), @() (spreadobject(currentresult, struct("expensive", {spreadarray(currentresult.("expensive"), struct("currentproduct", {currentproduct}))}))), @() (spreadobject(currentresult, struct("cheap", {spreadarray(currentresult.("cheap"), struct("currentproduct", {currentproduct}))})))), products, struct("expensive", {{}}, "cheap", {{}}));
sprint("grouped products: ", jsonstringify(productsgrouped, struct("pretty", {true})));
% grouped products: {
%     "expensive": [
%         {
%             "code": "pasta",
%             "price": 321
%         },
%         {
%             "code": "bubble_gum",
%             "price": 233
%         },
%         {
%             "code": "towel",
%             "price": 499
%         }
%     ],
%     "cheap": [
%         {
%             "code": "potato_chips",
%             "price": 5
%         }
%     ]
% }
